Update the metrics package

tiyash-basu-frequenz opened the pull request:
This PR refactors the metrics protobuf package to simplify and generalise it further.

tiyash-basu-frequenz committed:
Remove `rated_bounds` from `Metric` and `MetricAggregation`

`rated_bounds` contain mostly static information, which do not need to
be sent with every metric. These will now be available when listing
descriptions for components.

Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>

tiyash-basu-frequenz committed:
Remove `component_bounds` from `Metric` and `MetricAggregation`

The information contained in these bounds is redundant, since these are
supposed to be incorporated into the `system_inclusion_bounds` field.
Hence the `component_bounds` field has been removed from the messages.

Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>

tiyash-basu-frequenz committed:
Replace inclusion and exclusion bounds with a single array of bounds

This commit removes the `system_inclusion_bounds` and
`system_exclusion_bounds` fields from the `Metric` and
`MetricAggregation` messages. These fields have been replaced with a
single array of `Bounds` in both messages. This simplifies the message
definition, and removes the requirement of clients having to check if a
parameter is _not_ in a given pair of bounds. This also extends the
possibility of having more than 2 pairs bounds for a given metric.

Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>

tiyash-basu-frequenz committed:
Rename fields in `MetricAggregation` message

This commit renames the fields in the `MetricAggregation` message to
make them consistent with the `Metric` message.
The following renames have been done:
- `avg` -> `avg_value`
- `min` -> `min_value`
- `max` -> `max_value`

Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>

tiyash-basu-frequenz committed:
Add timestamps to metric messages

The `sampled_at` field has been added to both the `Metric` and
`MetricAggregation` messages. This makes it easier to use these messages
in a timeseries context, where each metric sample is associated with a
timestamp.

Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>

tiyash-basu-frequenz committed:
Rename `Metric` and `MetricAggregation` messages

This commit renames the `Metric` and `MetricAggregation` messages to
`SimpleMetricSample` and `AggregatedMetricSample` respectively.

Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>

tiyash-basu-frequenz committed:
Add a new message `MetricSampleVariant`

This message can encapsulate either a `SimpleMetricSample` or an
`AggregatedMetricSample`.

Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>

tiyash-basu-frequenz committed:
Add an enum containing valid metrics that can be streamed

Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>

thomas-nicolai-frequenz commented on proto/frequenz/api/common/v1/metrics.proto:
```diff
@@ -22,159 +24,222 @@ message Bounds {
+// List of supported metrics.
+enum Metric {
+  // Default value.
+  METRIC_UNSPECIFIED = 0;
```
Does it make sense to have the `METRIC_` prefix? Just wondering as the enum is already called Metric.

tiyash-basu-frequenz replied:
It is recommended to prefix protobuf enum variants with enum names. This is because if there are two or more enums in a scope, all their variants share the same scope (not the scope within the enum, but the outer scope to which the enums belong to).

So, we consistently prefix our protobuf enum variants with the enum names. Also, some programming languages support these by default. E.g., `METRIC_UNSPECIFIED` will appear just as `Unspecified` in rust.

thomas-nicolai-frequenz replied:
Got it. Thanks for clarifying!

tiyash-basu-frequenz committed:
Improve documentation style in metrics.proto

Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>

tiyash-basu-frequenz committed:
Add a message to represent metrics sampled from components

The newly added message
`frequenz.api.common.v1.components.data.ComponentData` represents
metrics, states, and errors sampled from components. Each such message
contains a list of `MetricSample` messages, each of which represents a
sampled metric along with its value. The message also contains a list of
`State` messages, each of which represents a component state and errors
sampled at a specific time.

Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>

thomas-nicolai-frequenz commented on proto/frequenz/api/common/v1/components.proto:
```diff
@@ -10,6 +10,217 @@ syntax = "proto3";
+//           }
+//         }
+//       },
+//       "state": "COMPONENT_STATE_CHARGING",
```
state is also an array now. the example needs adjusting.

tiyash-basu-frequenz replied:
Tom approved verbally.

thomas-nicolai-frequenz commented on proto/frequenz/api/common/v1/components.proto:
```diff
@@ -10,6 +10,217 @@ syntax = "proto3";
+    frequenz.api.common.v1.metrics.Metric metric = 1;
+
+    // The value of the sampled metric.
+    frequenz.api.common.v1.metrics.MetricSampleVariant value = 2;
```
I would suggest to call this sample and not value.

thomas-nicolai-frequenz replied:
And the prefix "frequenz.api.common.v1.metrics." can be remove as this is a self-reference.

tiyash-basu-frequenz replied:
The prefix cannot be removed, because the definitions of `Metric` and `MetricSampleVariant` are in a different file (and therefore under a different package hierarchy).

thomas-nicolai-frequenz replied:
oh right I overlooked that. Sorry my bad!

thomas-nicolai-frequenz replied:
> I would suggest to call this sample and not value.

This renaming depends on wether we keep the structure as suggest or change it to guarantee key uniqueness.

tiyash-basu-frequenz replied:
Agree about the renaming. `sample` is more meaningful than `value` here.

thomas-nicolai-frequenz commented on proto/frequenz/api/common/v1/components.proto:
```diff
@@ -10,6 +10,217 @@ syntax = "proto3";
+  // Representation of a sampled metric along with its value.
+  message MetricSample {
+    // The metric that was sampled.
+    frequenz.api.common.v1.metrics.Metric metric = 1;
```
The prefix "frequenz.api.common.v1.metrics." can be remove as this is a self-reference.

thomas-nicolai-frequenz commented on proto/frequenz/api/common/v1/components.proto:
```diff
@@ -10,6 +10,217 @@ syntax = "proto3";
+  // Default value.
+  COMPONENT_STATE_UNSPECIFIED = 0;
+
+  // The component is on.
```
Its off not on ;-)

thomas-nicolai-frequenz commented on proto/frequenz/api/common/v1/components.proto:
```diff
@@ -10,6 +10,217 @@ syntax = "proto3";
+  COMPONENT_STATE_UNAVAILABLE = 9;
+
+  // The component state is unknown.
+  COMPONENT_STATE_UNKNOWN = 10;
```
Is there a difference between UNKOWN and UNSPECIFIED? I feel its the same...

tiyash-basu-frequenz replied:
There is a difference, but maybe the wording needs to be improved.

`UNSPECIFIED` is used when a sender specifies no value for the enum. Protobuf serializers/deserializers will use this as a default in such cases.

`UNKOWN` is used when we get a state from a component, but we do not know what this state means (e.g., if this is not mentioned in the manual). So this would be specified, while being unknown to us. Our friend suggests `UNDEFINED` but not sure if it is an improvement.

thomas-nicolai-frequenz replied:
Yeah, that makes sense. I think it would be great to add this to the UNKOWN state to explain what the difference is to the UNSPECIFIED state.

tiyash-basu-frequenz replied:
Tom approved verbally.

thomas-nicolai-frequenz commented on proto/frequenz/api/common/v1/components.proto:
```diff
@@ -10,6 +10,217 @@ syntax = "proto3";
+// categories.
+enum ComponentError {
+  // Default value. No specific error is specified.
+  COMPONENT_ERROR_UNSPECIFIED = 0;
```
I would also have suggested to call this UKNOWN rather UNSPECIFIED but both is fine I guess. Just pointing it out...

thomas-nicolai-frequenz commented on proto/frequenz/api/common/v1/components.proto:
```diff
@@ -10,6 +10,217 @@ syntax = "proto3";
+  COMPONENT_ERROR_EV_PLUG_UNKNOWN = 24;
+
+  // Error indicating a non-standard electric vehicle (EV) consumer.
+  COMPONENT_ERROR_EV_CONSUMER_NON_STANDARD = 25;
```
I suggest to keep some gaps in the numbering between battery, ev etc.

thomas-nicolai-frequenz commented on proto/frequenz/api/common/v1/components.proto:
```diff
@@ -10,6 +10,217 @@ syntax = "proto3";
+  COMPONENT_ERROR_INTERNAL = 20;
+
+  // Error indicating electric vehicle (EV) cable lost.
+  COMPONENT_ERROR_EV_CABLE_LOST = 21;
```
I would put a gap in the numbering between common errors and component category specific ones. Maybe component specific ones should start at 100 or something? I mean there is plenty of "space".

thomas-nicolai-frequenz commented on proto/frequenz/api/common/v1/metrics.proto:
```diff
@@ -22,159 +24,223 @@ message Bounds {
+  METRIC_BATTERY_TEMPERATURE = 105;
+
+  // General inverter metrics.
+  METRIC_INVERTER_TEMPERATURE = 106;
```
again I would leave a gap between component specific metrics.

tiyash-basu-frequenz commented:
I have addressed the review comments @thomas-nicolai-frequenz . Ready for another look.

stefan-brus-frequenz commented on proto/frequenz/api/common/v1/metrics.proto:
```diff
@@ -22,159 +24,223 @@ message Bounds {
+  METRIC_DC_POWER = 3;
+
+  // General AC electricity metrics
+  METRIC_AC_FREQUENCY = 4;
```
Should there be more numeric "spacing" between the enum values here? E.g. using the values 11, 12, 13, 14 for one category of things, and then having the next one start at 20, in case you later find more "things" that belong to the previous category. I see that it's already done for other enums in this PR. I am not sure if it's necessary here though since this isn't really my domain, so up to you.

tiyash-basu-frequenz replied:
I added gaps between each group of electrical metrics.
